sklearn/metrics/_safe_euclidean_sparse.pyx

```diff
@@ -0,0 +1,69 @@
+#cython: language_level=3
+#cython: boundscheck=False, cdivision=True, wraparound=False
+
+
+import numpy as np
+cimport numpy as np
+from cython cimport floating
+from libc.math cimport fmax
+
+
+np.import_array()
+
+
+ctypedef fused INT:
+    np.int32_t
+    np.int64_t
+
+
+def _euclidean_sparse_dense_exact(floating[::1] X_data,
+                                  INT[::1] X_indices,
+                                  INT[::1] X_indptr,
```

Comment:
This assumes, CSR (not CSC) right? It could be worth indicating it in a comment.

Reply:
Right, CSR is forced by check_pairwise